src/v/container/include/container/chunked_vector.h

```diff
@@ -0,0 +1,296 @@
+        using value_type = typename std::conditional_t<Const, const T, T>;
+        using difference_type = std::ptrdiff_t;
+        using pointer = value_type*;
+        using reference = value_type&;
+
+        iter() = default;
+
+        reference operator*() const { return _vec->operator[](_index); }
+        pointer operator->() const { return &_vec->operator[](_index); }
+
+        iter& operator+=(ssize_t n) {
+            _index += n;
+            return *this;
+        }
+
+        iter& operator-=(ssize_t n) {
+            _index -= n;
+            return *this;
+        }
+
+        iter& operator++() {
+            ++_index;
+            return *this;
+        }
+
+        iter& operator--() {
+            --_index;
+            return *this;
+        }
+
+        iter operator++(int) {
+            auto tmp = *this;
+            ++*this;
+            return tmp;
+        }
+
+        iter operator--(int) {
+            auto tmp = *this;
+            --*this;
+            return tmp;
+        }
+
+        iter operator+(difference_type offset) { return iter{*this} += offset; }
+        iter operator-(difference_type offset) { return iter{*this} -= offset; }
+
+        bool operator==(const iter&) const = default;
+        auto operator<=>(const iter&) const = default;
+
+        friend ssize_t operator-(const iter& a, const iter& b) {
+            return a._index - b._index;
+        }
+
+    private:
+        friend class chunked_vector;
+        using vec_type = std::conditional_t<Const, const this_type, this_type>;
+
+        iter(vec_type* vec, size_t index)
+          : _index(index)
+          , _vec(vec) {}
+
+        size_t _index{};
+        vec_type* _vec{};
+    };
+
+    using const_iterator = iter<true>;
+    using iterator = iter<false>;
+
+    iterator begin() { return iterator(this, 0); }
+    iterator end() { return iterator(this, _size); }
+
+    const_iterator begin() const { return const_iterator(this, 0); }
+    const_iterator end() const { return const_iterator(this, _size); }
+
+    const_iterator cbegin() const { return const_iterator(this, 0); }
+    const_iterator cend() const { return const_iterator(this, _size); }
+
+    friend std::ostream& operator<<(std::ostream& os, const chunked_vector& v) {
+        os << "[";
+        for (auto& e : v) {
+            os << e << ",";
+        }
+        os << "]";
+        return os;
+    }
+
+private:
+    void maybe_add_capacity() {
+        if (_chunks.back().size() >= elements_per_chunk) {
+            std::vector<T> frag;
+            frag.reserve(elements_per_chunk);
+            _chunks.push_back(std::move(frag));
+        }
+    }
+
+private:
+    friend class chunked_vector_validator;
+    chunked_vector(const chunked_vector&) noexcept = default;
+
+    size_t _size{0};
+    // Creating a single vector to start means that `maybe_add_capacity` will
+    // grow like a normal vector for the first chunk.
+    std::vector<std::vector<T>> _chunks = {{}};
```

Comment:
This is clever but I think it's much better to leave `_chunks` empty for empty containers as empty containers are common and their size and construction performance matters (this is an important optimization for `std::vector` even though it might simplify the specification and implementation to do otherwise). Currently this will do an alloc even for empty containers.

Reply:
I had just pushed a change to use inlined vector here from absl so this does not happen, but point taken, maybe it is salvageable to just make a small modification to fragmented_vector